django/db/models/fields/related.py

```diff
@@ -613,6 +613,17 @@ def resolve_related_fields(self):
             )
             to_field = (self.remote_field.model._meta.pk if to_field_name is None
                         else self.remote_field.model._meta.get_field(to_field_name))
+            if to_field.model != self.remote_field.model._meta.concrete_model:
+                raise exceptions.FieldError(
+                    "ForeignKey '%s.%s' pointing to a non local remote field '%s.%s'." %
```

Comment:
This also stands for non-`ForeignKey` such as `ForeignObject` the way it is defined right now.

So I'd perform this check in `ForeignKey.resolve_related_fields` since it's the only one that has this restriction. That will allow `ForeignObject` to keep relying on this pattern and makes this patch slightly less backward incompatible.

That means that `ForeignObject(ChildModel, from_fields=('ref_id',), to_fields=('key',))` would still be allowed since `ForeignObject` doesn't require the enforcement of a database level foreign key and can span multiple tables.

Reply:
I agree with Simon, can you move this check to the `ForeignKey.resolve_related_fields()`?, e.g.
```python
def resolve_related_fields(self):
    related_fields = super().resolve_related_fields()
    for _, to_field in related_fields:
        if to_field and to_field.model != self.remote_field.model._meta.concrete_model:
            raise exceptions.FieldError(
                "ForeignKey '%s.%s' pointing to a non local remote field "
                "'%s.%s'." % (
                    self.model._meta.label,
                    self.name,
                    self.remote_field.model._meta.concrete_model._meta.label,
                    to_field.name,
                )
            )
    return related_fields
```